core/src/main/java/com/spotify/metrics/core/ReservoirWithTtl.java

```diff
@@ -105,50 +107,58 @@ public ReservoirWithTtl(
         this.now = now;
         this.ttlSeconds = ttlSeconds;
         this.bufferSize = ttlSeconds * minimumRate;
-        this.valueBuffer = EvictingQueue.create(bufferSize);
+        this.valueBuffer = new ConcurrentLinkedDeque<>();
+        this.valueBufferSize = new LongAdder();
     }
 
     @Override
     public int size() {
-        synchronized (this) {
-            purgeOld();
-            if (useInternalBuffer()) {
-                return valueBuffer.size();
-            }
+        purgeOld();
+        if (useInternalBuffer()) {
+            return Math.max(0, valueBufferSize.intValue());
         }
 
         return delegate.size();
     }
 
     @Override
     public void update(final long value) {
-        synchronized (this) {
-            valueBuffer.add(new ValueAndTimestamp(value, now.get()));
+        while (valueBufferSize.intValue() >= bufferSize) {
```

Comment:
I'm not sure this is safe (eg what happens when more threads arrive than bufferSize, are we relying on removeFirst to throw an exception to abort the decremented? might need a comment).

Reply:
I'm wondering if a circular queue with first/last atomic counters wouldn't be a better implementation of the queue since it's efficient to add and remove stuff into concurrently (ish) assuming you ignore overflow, you could perhaps move the overflow detection to the read path since that is not on the cirtical path and it might be ok to scan the buffer to see if it has overflowed (alternatively using some kind of lockless scheme to update start/end on overflow using https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/StampedLock.html but that could be trickier).

Reply:
now it's poll instead so now that is fixed. nvm.

Reply:
If we initialize the queue to size buffer (with fake old entries) this code becomes valueBuffer.pollFirst() ; valueBuffer.add() without any conditions.

Reply:
Great idea - fixed!